src/MongoDB.Driver.Core/Core/Authentication/DefaultAuthenticator.cs

```diff
@@ -14,76 +14,136 @@
     public class DefaultAuthenticator : IAuthenticator
     {
         // fields
         private readonly UsernamePasswordCredential _credential;
         private readonly IRandomStringGenerator _randomStringGenerator;
+        private readonly Lazy<BsonDocument> _saslSupportedMechanismsQuery;
 
         // constructors
         /// <summary>
-        /// Initializes a new instance of the <see cref="DefaultAuthenticator"/> class.
+        /// Initializes a new instance of the <see cref="DefaultAuthenticator" /> class.
         /// </summary>
         /// <param name="credential">The credential.</param>
         public DefaultAuthenticator(UsernamePasswordCredential credential)
             : this(credential, new DefaultRandomStringGenerator())
         {
         }
 
-        internal DefaultAuthenticator(UsernamePasswordCredential credential, IRandomStringGenerator randomStringGenerator)
+        internal DefaultAuthenticator(UsernamePasswordCredential credential,
+            IRandomStringGenerator randomStringGenerator)
         {
             _credential = Ensure.IsNotNull(credential, nameof(credential));
             _randomStringGenerator = Ensure.IsNotNull(randomStringGenerator, nameof(randomStringGenerator));
+            _saslSupportedMechanismsQuery = new Lazy<BsonDocument>(
+                ()=> CreateSaslSupportedMechsRequest(_credential.Source, _credential.Username));
         }
 
         // properties
-        /// <inheritdoc/>
-        public string Name
-        {
-            get { return "DEFAULT"; }
-        }
+        /// <inheritdoc />
+        public string Name => "DEFAULT";
+        
+        private BsonDocument SaslSupportedMechanismsQuery => _saslSupportedMechanismsQuery.Value;
 
         // methods
-        /// <inheritdoc/>
-        public void Authenticate(IConnection connection, ConnectionDescription description, CancellationToken cancellationToken)
+        /// <inheritdoc />
+        public void Authenticate(IConnection connection, ConnectionDescription description,
+            CancellationToken cancellationToken)
         {
             Ensure.IsNotNull(connection, nameof(connection));
             Ensure.IsNotNull(description, nameof(description));
 
-            var authenticator = CreateAuthenticator(description);
-            authenticator.Authenticate(connection, description, cancellationToken);
+            // if we've already piggybacked the initial isMaster request, we'll already have SaslSupportedMechs
+            // as part of the response. This check can be removed once multiple authenticators are no longer supported
+            if (description.IsMasterResult.SaslSupportedMechsReceived) 
+            {
+                CreateAuthenticator(connection, description).Authenticate(connection, description, cancellationToken);
+                return;
+            } // else, we're in a multi-authenticator--we did not piggyback, so we have to query for saslSupportedMechs 
+            
```

Comment:
One more note:

This additional mechanism negotiation step is only necessary if server version >= 4.0.

Reply:
Ack